### PR TITLE
Add parallelization for fetching images and pagination while fetching registry repositories

### PR DESCRIPTION
**Pull request by jatin-baweja**
Links to deepfence/enterprise-roadmap#2146
Links to #1994 

Changes proposed in this pull request:
- Add parallelization for fetching registry images
- Add pagination while fetching registry repositories

### CODE REVIEW

File: deepfence_server/pkg/registry/acr/client.go [hunk 1]
```diff
@@ -11,39 +12,103 @@ import (
...
-var client = &http.Client{Timeout: 10 * time.Second}
+const (
+	PerPageCount       = 100
+	ParallelImageFetch = 10
+)
+
+var (
+	client                = &http.Client{Timeout: 10 * time.Second}
+	parllelImageProcessor *tunny.Pool
+	queue                 chan []model.IngestedContainerImage
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/acr/client.go:23]: We need to make this a buffered channel. 
We can give this a buffer size of 100

File: deepfence_server/pkg/registry/acr/client.go [hunk 2]
```diff
@@ -11,39 +12,103 @@ import (
...
+)
+
+func init() {
+	parllelImageProcessor = tunny.NewFunc(ParallelImageFetch, fetchImageWithTags)
+	queue = make(chan []model.IngestedContainerImage)
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/acr/client.go:28]: We need to make this a buffered channel.
We can give this a buffer size of 100

File: deepfence_server/pkg/registry/acr/client.go [hunk 3]
```diff
@@ -11,39 +12,103 @@ import (
...
+			Password:   password,
+			NameSpace:  namespace,
+			Repository: repo,
+		}
+		go parllelImageProcessor.Process(r)
+	}
+	for _, _ = range repos {
+		select {
+		case t := <-queue:
+			images = append(images, t...)
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/acr/client.go:61]: Do we need to check if the `t` is empty before we do the append?

File: deepfence_server/pkg/registry/acr/client.go [hunk 4]
```diff
@@ -11,39 +12,103 @@ import (
...
 	return images, nil
 }
 
-func listCatalogRegistryV2(url, namespace, userName,
-	password string) ([]string, error) {
+func fetchImageWithTags(rInterface interface{}) interface{} {
+	r, ok := rInterface.(*RepoDetails)
+	if !ok {
+		log.Error().Msg("Error processing repo details")
+		queue <- []model.IngestedContainerImage{}
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/acr/client.go:72]: We should create a variable `images` at top 
```
images := model.IngestedContainerImage{}
```
And use the `defer queue <- images` 
Reason being, if later we forget to add the image slice to the channel, the consumer side will be blocked forever.

File: deepfence_server/pkg/registry/dockerprivate/client.go [hunk 5]
```diff
@@ -12,47 +13,109 @@ import (
...
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-	},
+const (
+	PerPageCount       = 100
+	ParallelImageFetch = 10
+)
+
+func init() {
+	parllelImageProcessor = tunny.NewFunc(ParallelImageFetch, fetchImageWithTags)
+	queue = make(chan []model.IngestedContainerImage)
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/dockerprivate/client.go:23]: Same comment as above. We can make this a buffered channel.

File: deepfence_server/pkg/registry/acr/client.go [hunk 6]
```diff
@@ -11,39 +12,103 @@ import (
 	"github.com/deepfence/ThreatMapper/deepfence_utils/log"
 )
 
-var client = &http.Client{Timeout: 10 * time.Second}
+const (
+	PerPageCount       = 100
+	ParallelImageFetch = 10
```
Comment (varunsharma0286) [inline deepfence_server/pkg/registry/acr/client.go:17]: Should we keep a low and high threshold for the number of workers?
Reason being, if we have less number of repositories, we can work with less number of workers. 

In the init, we create the number of workers = low threshold. 
Lets say we want to allocate 10 repositories per worker. So we can calculate the number of workers using `num_repo/10`. If this count is > 10, we limit the number to high threshold of 10.

**Review by varunsharma0286 — APPROVED** [review]
+1

**Review by ibreakthecloud — APPROVED** [review]

**Pull request merged**